fdbserver/storageserver.actor.cpp

```diff
@@ -1090,100 +1144,105 @@ ACTOR Future<Void> getValueQ( StorageServer* data, GetValueRequest req ) {
-					}
-					if(BUGGIFY) {
-						// Simulate a trigger on the watch that results in the loop going around without the value changing
-						watchFuture = watchFuture || delay(deterministicRandom()->random01());
-					}
-					wait(watchFuture);
-					--data->numWatches;
-					data->watchBytes -= (req.key.expectedSize() + req.value.expectedSize() + WATCH_OVERHEAD_BYTES);
-				} catch( Error &e ) {
-					--data->numWatches;
-					data->watchBytes -= (req.key.expectedSize() + req.value.expectedSize() + WATCH_OVERHEAD_BYTES);
-					throw;
+			data->watchBytes +=
+			    (metadata->key.expectedSize() + metadata->value.expectedSize() + key.expectedSize() +
+			     sizeof(Reference<ServerWatchMetadata>) + sizeof(ServerWatchMetadata) + WATCH_OVERHEAD_WATCHIMPL);
+			try {
+				if(latest < minVersion) {
+					// If the version we read is less than minVersion, then we may fail to be notified of any changes that occur up to or including minVersion
+					// To prevent that, we'll check the key again once the version reaches our minVersion
+					watchFuture = watchFuture || data->version.whenAtLeast(minVersion);
 				}
-			} catch( Error &e ) {
-				if( e.code() != error_code_transaction_too_old ) {
-					throw;
+				if(BUGGIFY) {
+					// Simulate a trigger on the watch that results in the loop going around without the value changing
+					watchFuture = watchFuture || delay(deterministicRandom()->random01());
 				}
-
-				TEST(true); // Reading a watched key failed with transaction_too_old
+				wait(watchFuture);
+				data->watchBytes -=
+				    (metadata->key.expectedSize() + metadata->value.expectedSize() + key.expectedSize() +
+				     sizeof(Reference<ServerWatchMetadata>) + sizeof(ServerWatchMetadata) + WATCH_OVERHEAD_WATCHIMPL);
+			} catch( Error &e ) {
+				data->watchBytes -=
+				    (metadata->key.expectedSize() + metadata->value.expectedSize() + key.expectedSize() +
+				     sizeof(Reference<ServerWatchMetadata>) + sizeof(ServerWatchMetadata) + WATCH_OVERHEAD_WATCHIMPL);
+				throw;
+			}
+		} catch( Error &e ) {
+			if( e.code() != error_code_transaction_too_old ) {
+				throw e;
 			}
 
-			watchFuture = data->watches.onChange(req.key);
-			wait(data->version.whenAtLeast(data->data().latestVersion));
+			TEST(true); // Reading a watched key failed with transaction_too_old
 		}
-	} catch (Error& e) {
-		if(!canReplyWith(e))
-			throw;
-		data->sendErrorWithPenalty(req.reply, e, data->getPenalty());
+
+		watchFuture = data->watches.onChange(metadata->key);
+		wait(data->version.whenAtLeast(data->data().latestVersion));
+	}
+}
+
+void checkCancelWatchImpl( StorageServer* data, WatchValueRequest req ) {
+	Reference<ServerWatchMetadata> metadata = data->getWatchMetadata(req.key.contents());
+	if(metadata.isValid() && metadata->versionPromise.getFutureReferenceCount() == 1) {
+		// last watch timed out so cancel watch_impl and delete key from the map
+		data->deleteWatchMetadata(req.key.contents());
+		metadata->watch_impl.cancel();
```

Comment:
```suggestion
```
This is unnecessary. When you delete the `metadata` object it will cancel the future.

Reply:
I believe this call is necessary because `watch_impl` actor also contains a reference to the [WatchMetadata](https://github.com/apple/foundationdb/blob/5462d52c8b2ecdd9027fed8376e5c09a01e71b40/fdbserver/storageserver.actor.cpp#L1165). So if we don't cancel it here then deleting from the map has no impact because we will still have a reference count of 1 causing us to never cancel this actor.

Reply:
Ah I see now -- this is a bit of a cyclic dependency. But I guess in this case this is ok.